### PR TITLE
CASSANDRA-18692 Fix bulk writes with Buffered RowBufferMode

### PR DESCRIPTION
**Pull request by frankgh**
When setting Buffered RowBufferMode as part of the `WriterOption`s, `org.apache.cassandra.spark.bulkwriter.RecordWriter` ignores that configuration and instead uses the batch size to determine when to finalize an SSTable and start writing a new SSTable, if more rows are available.

In this commit, we fix `org.apache.cassandra.spark.bulkwriter.RecordWriter#checkBatchSize` to take into account the configured `RowBufferMode`. And in specific to the case of the `UNBUFFERED` RowBufferMode, we check then the batchSize of the SSTable during writes, and for the case of `BUFFERED` that check will take no effect.

Co-authored-by: Doug Rohrer <doug@therohrers.org>

### CODE REVIEW

**Comment by dineshjoshi** [comment]
+1 LGTM.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java [hunk 1]
```diff
@@ -114,16 +115,21 @@ public StreamResult write(Iterator<Tuple2<DecoratedKey, Object[]>> sourceIterato
                 checkBatchSize(streamSession, partitionId, job);
             }
 
-            if (batchSize != 0)
+            if (sstableWriter != null)
             {
-               finalizeSSTable(streamSession, partitionId, sstableWriter, batchNumber, batchSize);
+                finalizeSSTable(streamSession, partitionId, sstableWriter, batchNumber, batchSize);
```
Comment (yifan-c) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:120]: nit: reset `sstableWriter` to `null` after `finalizeSSTable`

File: cassandra-four-zero/src/main/java/org/apache/cassandra/bridge/SSTableWriterImplementation.java [hunk 2]
```diff
@@ -80,4 +79,28 @@ public void close() throws IOException
...
+                                                           .forTable(createStatement)
+                                                           .withPartitioner(cassPartitioner)
+                                                           .using(insertStatement);
+        if (rowBufferMode == RowBufferMode.UNBUFFERED)
+        {
+            builder.sorted();
+        }
+        else if (rowBufferMode == RowBufferMode.BUFFERED)
+        {
+            builder.withBufferSizeInMB(bufferSizeMB);
```
Comment (yifan-c) [inline cassandra-four-zero/src/main/java/org/apache/cassandra/bridge/SSTableWriterImplementation.java:102]: nit: maybe decide a valid value range for `bufferSizeMB` and validate. `CQLSSTableWriter` accepts whatever int value. 
For the upper bound, I think 1/2 of the `spark.executor.memory` is a good limit.

**Review by yifan-c — COMMENTED** [review]
some nits. Looks good to me.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java (hunk 1 above)
Comment (JeetKunDoug) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java:120]: Given this is essentially the last step before we're done with the sstablewriter, we don't really need to set the sstablewriter to null here - this RecordWriter instance should be collected as soon as the upload/commits finish, and we close the sstablewriter on finalize so we should be good w/o nulling it out.

File: cassandra-four-zero/src/main/java/org/apache/cassandra/bridge/SSTableWriterImplementation.java (hunk 2 above)
Comment (JeetKunDoug) [reply cassandra-four-zero/src/main/java/org/apache/cassandra/bridge/SSTableWriterImplementation.java:102]: I think, for now, we leave this just a configuration option... Validating it given the Spark environment and picking a reasonable upper-bound would take some experimentation (and would likely involve more than just the executor.memory setting, as there are other config settings that deal w/ memory like memory overhead).

**Pull request closed without merging**